File: product_sdk/create_product_request.go
package productsdk

import (
	"bytes"
	"encoding/json"
	"io"

	"github.com/go-playground/validator/v10"
)

type CreateProductRequest struct {
	Name        string  `json:"name" validate:"required"`
	Description string  `json:"description" validate:"required"`
	Price       float64 `json:"price" validate:"required"`
	Quantity    int     `json:"quantity" validate:"required,gt=0"`
}

func (c *CreateProductRequest) ToJson() io.Reader {
	buffer := new(bytes.Buffer)
	json.NewEncoder(buffer).Encode(c)

	return buffer
}

func (c *CreateProductRequest) IsValid() bool {
	Validate = validator.New()
	err := Validate.Struct(c)

	return err == nil
}

File: frontend/shoppie_admin/lib/features/products/models/create_product_request.dart
import 'package:json_annotation/json_annotation.dart';

part "create_product_request.g.dart";

@JsonSerializable()
class CreateProductRequest {
  final String name;
  final String description;
  final double price;
  final int quantity;

  CreateProductRequest({
    required this.description,
    required this.name,
    required this.price,
    required this.quantity,
  });

  factory CreateProductRequest.fromJson(Map<String, dynamic> json) =>
      _$CreateProductRequestFromJson(json);

  Map<String, dynamic> toJson() => _$CreateProductRequestToJson(this);
}
